scripts/triggers-end-to-end-tests/tests.ts

```diff
@@ -0,0 +1,271 @@
+import { expect } from "chai";
+import * as admin from "firebase-admin";
+import { Firestore } from "@google-cloud/firestore";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+
+import { CLIProcess } from "./cli";
+import { TriggerEndToEndTest } from "./framework";
+
+const FIREBASE_PROJECT = process.env.FBTOOLS_TARGET_PROJECT || "";
+
+const ALL_EMULATORS_STARTED_LOG = "All emulators started, it is now safe to connect.";
+
+/*
+ * Various delays that are needed because this test spawns
+ * parallel emulator subprocesses.
+ */
+const TEST_SETUP_TIMEOUT = 60000;
+const EMULATORS_WRITE_DELAY_MS = 5000;
+const EMULATORS_SHUTDOWN_DELAY_MS = 5000;
+const EMULATOR_TEST_TIMEOUT = EMULATORS_WRITE_DELAY_MS * 2;
+
+/*
+ * Realtime Database and Firestore documents we used to verify
+ * bidirectional communication between the two via cloud functions.
+ */
+const FIRESTORE_COMPLETION_MARKER = "test/done_from_firestore";
+const DATABASE_COMPLETION_MARKER = "test/done_from_database";
+
+function readConfig(): unknown {
+  const filename = path.join(__dirname, "firebase.json");
+  const data = fs.readFileSync(filename, "utf8");
+  return JSON.parse(data);
+}
+
+describe("database and firestore emulator function triggers", () => {
```

Comment:
I am going to assume this file is pretty much unchanged, let me know if I need to take a closer look at it.

Reply:
The only thing that changed is the `after` functions, since it now tries to clean up some connections. But other than that it's the same.